doc/plaintextimport.1
.\" Man page for plaintextimport
.\" Use the following command to view man page:
.\"
.\"  tbl plaintextimport.1 | nroff -man | less
.\"
.TH PLAINTEXTIMPORT 1 "August 9, 2006"
.SH NAME
plaintextimport \- A tool to convert plain text to Emdros MQL
.SH SYNOPSIS
\fBplaintextimport\fR [ options ] [input_filename ...]
.br
.SH DESCRIPTION
\fBplaintextimport\fR is a command-line tool to convert plain text
to Emdros MQL for later importing into Emdros.

.SH OPTIONS
\fBplaintextimport\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help, then quit
.TP
.BI \-V\ ,\ \-\-version
show version, then quit
.TP
.BI \-s\ ,\ \-\-schema
show MQL schema on stdout, then quit (can be used with -d)
.TP
.BI \-d\ ,\ \-\-dbname \ dbname 
set database name. If used with -s, the string "CREATE DATABASE
'dbname' GO USE DATABASE 'dbname' GO" will be issued before the schema.  If used when importing, "USE DATABASE 'dbname' GO" will be issued before anything else.
.TP
.BI \-o\ ,\ \-\-output \ filename
dump to file filename. The default is "-", which means "standard
output".
.TP
.BI \-\-start-monad \ monad
The start monad to use. Must be >= 1.  The default is 1.
.TP
.BI \-\-start-id_d \ id_d
The start id_d to use. Must be >= 1.  The default is 1.
.TP

.SH OPERATION

\fBplaintextimport\fR reads plain text converts the text to MQL
statements for later importing into Emdros.

The filenames given after the options on the command line are
interpreted as if each of them contains one document, each containing
a whitespace-separated string of words.  If no filenames are given,
the input is read from stdin.

If no \fB-o\fR switch is given, the output is printed on stdout.

If an error occurs, the string "FAILURE" or the string "ERROR" is
printed on stderr, along with an error message.

If no error occurs, a string of the form "SUCCESS: next_monad is X
next_id_d is Y" is printed on stderr, where X and Y are positive
integers denoting the next monad and the next id_d to be used by the
next invocation of the program, respectively.  This is useful if
you've got several directories' worth of documents to import.

.SH SCHEMA

The schema can be seen by giving the program \fB-s\fR switch, with an
optional \fB-d\fR switch.

A "Document" corresponds to one top-level file.

A "Word" is a single whitespace-separated "string" in a document.


.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Could not open file"
.TP
.BR 5 " Database error"
.TP
.BR 6 " Compiler error (internal error)"
.SH AUTHORS
Copyright
.Cr
2001-2006 by Ulrik Petersen (ulrikp@users.sourceforge.net).  Note that
this software is distributed under the GNU GPL.  See the sources for
details.


NEWS
This file is for release-notes for Emdros.

- *** Version 3.5.0 *** 

Release notes for Emdros 3.5.0
==============================

The focus-areas of this release include:

- Enhancements to the core MQL
- Bugfixes to the core MQL.
- Enhancements to the render library
- Enhancements to the Full Text Search subsystem
- Enhancements to SWIG bindings
- Enhancements to the build system

Please consult the ChangeLog (going back to version 3.4.1.pre01) for
the specifics.  You can also consult the MQL Programmer's Guide for
the enhancements to the core MQL, and the Programmer's Reference Guide
for the enhancements to the Render library and the FTS subsystem.

Briefly, the core MQL has been enhanced as follows:

- SELECT ALL OBJECTS has a new WITH MAX RANGE clause, helping to limit
  otherwise runaway queries such as [Word]..[Word].

- SELECT ALL OBJECTS now sports an [UnorderedGroup] block.

- GET AGGREGATE FEAUTURES is a new statement, similar to SQL's MIN(),
  MAX(), SUM(), and COUNT() aggregate functions.


- *** Version 3.4.0 *** 

Release notes for Emdros 3.4.0
==============================

The focus-areas of this release include:

- Support for iOS and Android
- A smoother user experience for the GUI applications
- Enhancements to the MQL query language
- Documentation improvements
- Build improvements on all platforms
- Speed improvements


Binaries for Windows(R) and Mac OS X(R)
---------------------------------------

As usual, binaries are available for Mac OS X and Windows(R).

The Windows binaries are available in both x86 and x64 versions.  They
have built-in support for MySQL, SQLite 2, and SQLite 3.  The x86
binaries are built with Visual Studio Express 2010.  The x64 binaries
are built with Windows SDK 6.0A.

The Mac OS X binaries are Universal binaries (i386 and x86_64) running
on Mac OS X 10.6 (Snow Leopard), 10.7 (Lion), and 10.8 (Mountain
Lion), 10.9 (Mavericks), and possibly later versions of Mac OS X.
They do not have support for either MySQL or PostgreSQL; Only SQLite 2
and SQLite 3 are supported in the Mac OS X binaries.  You might be
able to compile the sources with support for MySQL yourself, and
possibly also PostgreSQL.

Linux binaries
--------------

For this release, we do not provide Linux binaries.  Instead, we have
ensured that it is easy to build .rpms and .debs from the sources.


To build an RPM or .deb, first install the dependencies mentioned in
the text file called "INSTALL" in the root of the sources.


A) To build an RPM:

rpmbuild -ta emdros-3.4.0.tar.gz


B) To build a .deb:

1) cp /path/to/Emdros/sources/emdros-3.4.0.tar.gz .

2) tar xfvz emdros-3.4.0.tar.gz

3) cd emdros-3.4.0

4) dpkg-buildpackage -rfakeroot

This will build the .deb in ../.


Changes in more detail
----------------------

Release 3.4.0 has the following changes over 3.3.0:

- Build changes:

  - Support for compilation on Android and iOS were added.

  - Non-Unicode (ANSI) builds of the programs are no longer available
    on Windows.

  - On Linux / Unix / Mac OS X, wx/libemdrosguiu is now a static
    library.

  - Out-of-the-box support for building on Debian / Ubuntu /
    derivatives was added:

	$ tar xfzv emdros-x.y.z.tar.gz
	$ cd emdros-x.y.z
	$ dpkg-buildpackage -rfakeroot
	$ cd .. # The .deb package is in this directory
	$ sudo dpkg -i emdros_*.deb # This should install it.

  - The .pro build files for Qt have been updated to Qt 4.8.0, which
    entails dropping support for everything below Mac OS X 10.6 on Qt.

  - It is now possible to do Win64 builds on Windows.  See
    the very top of win32\config.mak for how to do it.

  - It is now possible to build Emdros on Windows with MinGW32,
    including support for wxWidgets.

  - Add a test suite which is run with "make test" in the root
    directory of the sources.

  - Much better configure.ac support for wxWidgets 2.8, 2.9, and 3.0,
    especially on Mac OS X.


- Harvest library changes:

  - Bugs were fixed
  - Speeds were increased
  - The template language was enhanced (see the docs)
  - A new Full Text Search engine, FTS2, was added.

- PHP was added as a supported SWIG binding.

- More support for feature-type "SET OF MONADS" was added:

  - Bugs were fixed
  - In topographic queries, it is now possible to say:

    SELECT ALL OBJECTS WHERE 
    [Clause 
        [Phrase
           // Could also be universe instead of substrate.
           // Could also be part_of instead of overlap.
           // part_of(monads, substrate) is still the default.
           overlap(mymonadsetfeature, substrate)
        ]
    ]

- The GET OBJECTS HAVING MONADS IN statement now supports using a
  different monad set than the default "monads" feature:

  GET OBJECTS HAVING MONADS IN
  { 2-40201, 40203-50000 }
  USING MONAD FEATURE described_SOM
  [Describer]
  GO

- Assorted speed increases in the EMdF / MQL libraries.

- The Penn Treebank converter output was improved.

- SQLite 3 was updated to version 3.8.4.3.

- Fixed a few documentation glitches


Enjoy!


Ulrik Sandborg-Petersen


- *** Version 3.3.0 *** 

As usual, binaries are available for Mac OS X, Windows(R), and Fedora
(15).  

The Windows binaries have support for MySQL, SQLite 2, and SQLite 3.
They are built with Visual Studio Express 2008.

The Mac OS X binaries are Universal binaries running on Mac OS X 10.4
(Tiger), 10.5 (Leopard), and 10.6 (Snow Leopard).  They do not have
support for either MySQL or PostgreSQL; Only SQLite 2 and SQLite 3 are
supported in the Mac OS X binaries.  You can compile the sources with
support for MySQL yourself, though, and possibly also PostgreSQL.

The Fedora binaries come with support for PostgreSQL, MySQL, SQLite 2,
and SQLite 3.

Release 3.3.0 has the following changes over 3.2.0:


src/plaintextimport.cpp
/*
 * plaintextimport.cpp
 *
 * A front-end to Emdros for importing plain text
 * Created: 8/8-2006 (August 8, 2006)
 * Last update: 1/20-2007
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : Could not open file
 * 5 : Database error
 * 6 : Compiler error
 *
 */
/************************************************************************
 *
 *   Emdros - the database engine for analyzed or annotated text
 *   Copyright (C) 2001-2007  Ulrik Petersen
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation, license version 2.  
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *   General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
 *   02111-1307 USA
 *
 *
 *   Special exception
 *   =================
 * 
 *   In addition, as a special exception, Ulrik Petersen, the
 *   copyright holder of Emdros, gives permission to link Emdros, in
 *   whole or in part, with the libraries which are normally
 *   distributed with:
 *   
 *   a) Sun's Java platform,
 *   b) Python, 
 *   c) Jython,
 *   d) Ruby, and/or 
 *   e) Perl 
 *   f) PostgreSQL
 *   g) OpenSSL
 *
 *   (or with modified versions of these), and to distribute linked
 *   combinations including both Emdros, in whole or in part, and one
 *   or more of the libraries normally distributed with (a)-(g) above.
 *
 *   Please note: This gives you special rights concerning the
 *   libraries which normally accompany the above pieces of software.
 *   It gives you no special rights concerning software that you write
 *   yourself.  You must obey the GNU General Public License in all
 *   respects for all of the code used other than the libraries
 *   normally distributed with (a)-(g) above.
 *
 *   If you modify this file, you may extend this exception to your
 *   version of the file, but you are not obligated to do so. If you
 *   do not wish to do so, delete this exception statement from your
 *   version.
 *
 *
 *   Other licensing forms
 *   =====================
 *
 *   If you wish to negotiate commercial licensing, please contact
 *   Ulrik Petersen at ulrikp[at]users.sourceforge.net.
 *
 *   Licensing can also be negotiated if your organization is an
 *   educational, non-profit, charity, missionary or similar
 *   organization.
 *
 *
 *   Website
 *   =======
 *
 *   Emdros has a website here:
 *
 *   http://emdros.org
 *
 *
 *
 **************************************************************************/


#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc


#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <string_func.h>
#include <emdros-lconfig.h>
#include <opt.h>
#include <plaintextimporter.h>




int exec_list(const std::list<std::string>& input_filename_list, std::ostream* pOutStream, const std::string& dbname, monad_m start_monad, id_d_t start_id_d, bool bPrintSchema)
{
	int result = 0;

	PlainTextImporter *pImporter = new PlainTextImporter(start_monad, start_id_d);

	if (bPrintSchema) {
		if (dbname != "") {
			std::cout << "CREATE DATABASE '" << dbname << "' GO\n\n"
				  << "USE DATABASE '" << dbname << "' GO\n\n";
		}
		
		pImporter->putSchema(&std::cout);

		delete pImporter;

		return 0;
	}


	bool bResult = false;
	if (input_filename_list.empty()) {
		bResult = pImporter->readDocumentFromStream(&std::cin);
	} else {
		bResult = pImporter->readDocumentsFromFiles(input_filename_list);
	}

	if (!bResult) {
		std::cerr << "FAILURE: Error occurred reading documents." << std::endl;
		result = 5;
	} else {
		if (dbname != "") {
			(*pOutStream) << "USE DATABASE '" << dbname << "' GO\n\n";
		}

		pImporter->putCorpusAsMQL(pOutStream);

		std::cerr << "SUCCESS: next_monad is " << pImporter->getNextMonad() << "   next_id_d is " << pImporter->getNextID_D() << std::endl;
	}

	delete pImporter;

	return result;
}

bool open_file_for_writing(std::ofstream& fout, const std::string& filename)
{
	bool bResult = false;
	fout.open(filename.c_str());
	if (!fout) {
		std::cerr << "FAILURE: Could not open file " << filename << " for writing." << std::endl;
		std::cerr << "Action aborted." << std::endl;
		bResult = false;
	} else {
		bResult = true;
	}

	return bResult;
}

bool open_file_for_reading(std::ifstream& fin, const std::string& filename)
{
	bool bResult = false;
	fin.open(filename.c_str());
	if (!fin) {
		std::cerr << "FAILURE: Could not open file " << filename << " for reading." << std::endl;
		std::cerr << "Action aborted." << std::endl;
		bResult = false;
	} else {
		bResult = true;
	}

	return bResult;
}

void show_version(std::ostream& ostr)
{
	ostr << "plaintextimport from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   plaintextimport [options] [inputfile ...]" << std::endl;
	ostr << "   plaintextimport (--help|-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	ostr << "   --help               Show this help" << std::endl;
	ostr << "   -V , --version       Show version" << std::endl;
	ostr << "   -s , --schema        Print schema and quit." << std::endl;
	ostr << "   -d , --dbname db     database name (to use in MQL output)" << std::endl;
	ostr << "   -o , --output file   Dump database to filename file (\"-\" means stdout)" << std::endl;
	ostr << "   --start-monad monad  Start monad (default 1)" << std::endl;
	ostr << "   --start-id_d id_d    Start id_d (default 1)" << std::endl;
	ostr << "DEFAULTS:" << std::endl;
	ostr << "   -o -                 (Use stdout)" << std::endl;
	ostr << "   --start-monad 1" << std::endl;
	ostr << "   --start-id_d  1" << std::endl;
	ostr << "   If no filenames are given, a document is read from stdin." << std::endl;
	ostr << "   By default, no -d switch is assumed." << std::endl;
}

int main(int argc, char* argv[])
{
	// Set defaults
	std::string db;
	std::string output_filename("-");
	std::list<std::string> input_filename_list;
	bool bPrintSchema = false;
	monad_m start_monad = 1;
	id_d_t start_id_d = 1;
	std::string dummy_hostname, dummy_user, dummy_password;
	eCharsets dummy_charset;
	eBackendKind dummy_backend_kind;

	addStandardArguments(false // Do not add -e | --encoding
			     );

	addOption("-s", "--schema");

	addOption("-d", "--dbname", true, 
		  "", // Default value
		  "ERROR: -d and --dbname must have a database-name as their argument.\n"
		  "       example: -d book1\n");

	addOption("-o", "--output", true, 
		  "-", // Default value
		  "ERROR: -o and --output must have a filename (or \"-\" for stdout)\n"
		  "       as their argument.\n"
		  "       example: -o output.xml\n");


	addOption("--start-monad", "--start-monad", true, 
		  "1", // Default value
		  "ERROR: --start-monad must have a monad as its argument\n"
		  "       example: --start-monad 1\n");

	addOption("--start-id_d", "--start-id_d", true, 
		  "1", // Default value
		  "ERROR: --start-id_d must have an id_d as its argument\n"
		  "       example: --start-id_d 1\n");


	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  dummy_hostname,
					  dummy_user,
					  dummy_password,
					  dummy_backend_kind,
					  dummy_charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		// --start-monad
		std::string start_string;
		if (getArgumentValue("--start-monad", start_string)) {
			if (string_is_number(start_string)) {
				start_monad = string2long(start_string);
			} else {
				print_usage(std::cerr);
				std::cerr << "ERROR: The argument of --start-monad must be a number, and not '" << start_string << "'." << std::endl;
				return 1;
			}
		}
		if (start_monad < 1) {
			print_usage(std::cerr);
			std::cerr << "ERROR: The argument of --start-monad must be >= 1." << std::endl;
			return 1;
		}

		// --start-id_d
		if (getArgumentValue("--start-id_d", start_string)) {
			if (string_is_number(start_string)) {
				start_id_d = string2long(start_string);
			} else {
				print_usage(std::cerr);
				std::cerr << "ERROR: The argument of --start-id_d must be a number, and not '" << start_string << "'." << std::endl;
				return 1;
			}
		}
		if (start_id_d < 1) {
			print_usage(std::cerr);
			std::cerr << "ERROR: The argument of --start-id_d must be >= 1." << std::endl;
			return 1;
		}

		// --dbname | -d
		getArgumentValue("-d", db);

		// Output
		getArgumentValue("-o", output_filename);

		// -s | --schema
		if (getArgumentPresent("-s")) {
			bPrintSchema = true;
		}

		// Any and all surplus arguments are input filenames...
		input_filename_list = surplus_arguments;
	}


	int nResult = 3;
	bool bContinue = true;
	try {
		std::ostream *pOut = 0;
		std::ofstream fout;
		if (output_filename == "-") {
			pOut = &std::cout;
		} else {
			if (!open_file_for_writing(fout, output_filename)) {
				nResult = 4;
				bContinue = false;
			} else {
				pOut = &fout;
			}
		}

		if (bContinue) {
			nResult = exec_list(input_filename_list, pOut, db, start_monad, start_id_d, bPrintSchema);
		}
	} catch (EMdFNULLValueException e) {
		std::cerr << "ERROR: EMdFNULLValueException (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBDBError e) {
		std::cerr << "ERROR: EMdFDBDBError (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBException e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		/*
		std::cerr << pEE->getDBError() << std::endl; 
		std::cerr << pEE->getCompilerError() << std::endl;
		*/
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableColumnException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableRowException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	}

	// Clean up
	//	delete pEE;
	return nResult;
}


